test_conformance/extensions/cl_khr_command_buffer/command_buffer_event_sync.cpp

```diff
@@ -0,0 +1,1068 @@
+    {
+        log_error("EventMode::RET_COMBUF_WAIT_FOR_REGULAR test with out of "
+                  "order queue failed\n");
+        return TEST_FAIL;
+    }
+    return TEST_PASS;
+}
+
+int test_wait_for_sec_queue_event(cl_device_id device, cl_context context,
+                                  cl_command_queue queue, int num_elements)
+{
+    if (TEST_PASS
+        != MakeAndRunTest<CommandBufferEventSync<
+            EventMode::RET_WAIT_FOR_SEC_QUEUE_EVENT, false>>(
+            device, context, queue, num_elements))
+    {
+        log_error("EventMode::RET_WAIT_FOR_SEC_QUEUE_EVENT test with in order "
+                  "queue failed\n");
+        return TEST_FAIL;
+    }
+
+    if (TEST_PASS
+        != MakeAndRunTest<CommandBufferEventSync<
+            EventMode::RET_WAIT_FOR_SEC_QUEUE_EVENT, true>>(
+            device, context, queue, num_elements))
+    {
+        log_error("EventMode::RET_WAIT_FOR_SEC_QUEUE_EVENT test with out of "
+                  "order queue failed\n");
+        return TEST_FAIL;
+    }
+    return TEST_PASS;
+}
+
+//--------------------------------------------------------------------------
+// user-events test cases
+
+int test_user_event_wait(cl_device_id device, cl_context context,
+                         cl_command_queue queue, int num_elements)
+{
+    if (TEST_PASS
+        != MakeAndRunTest<
+            CommandBufferEventSync<EventMode::USER_EVENT_WAIT, false>>(
+            device, context, queue, num_elements))
+    {
+        log_error(
+            "EventMode::USER_EVENT_WAIT test with in order queue failed\n");
+        return TEST_FAIL;
+    }
+
+    if (TEST_PASS
+        != MakeAndRunTest<
+            CommandBufferEventSync<EventMode::USER_EVENT_WAIT, true>>(
+            device, context, queue, num_elements))
+    {
+        log_error(
+            "EventMode::USER_EVENT_WAIT test with out of order queue failed\n");
+        return TEST_FAIL;
+    }
+    return TEST_PASS;
+}
+
+int test_user_events_wait(cl_device_id device, cl_context context,
+                          cl_command_queue queue, int num_elements)
+{
+    if (TEST_PASS
+        != MakeAndRunTest<
+            CommandBufferEventSync<EventMode::USER_EVENTS_WAIT, false>>(
+            device, context, queue, num_elements))
+    {
+        log_error(
+            "EventMode::USER_EVENTS_WAIT test with in order queue failed\n");
+        return TEST_FAIL;
+    }
+
+    if (TEST_PASS
+        != MakeAndRunTest<
+            CommandBufferEventSync<EventMode::USER_EVENTS_WAIT, true>>(
+            device, context, queue, num_elements))
+    {
+        log_error("EventMode::USER_EVENTS_WAIT test with out of order queue "
+                  "failed\n");
+        return TEST_FAIL;
+    }
+    return TEST_PASS;
+}
+
+int test_user_event_callback(cl_device_id device, cl_context context,
+                             cl_command_queue queue, int num_elements)
+{
+    if (TEST_PASS
+        != MakeAndRunTest<
```

Comment:
I don't think this pattern is handling test skips correctly. If I try these tests against and implementation which doesn't support the command-buffer extension, then it fails rather than skips

```
regular_wait_for_command_buffer...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_REGULAR_WAIT_FOR_COMBUF test with in order queue failed
regular_wait_for_command_buffer FAILED
command_buffer_wait_for_command_buffer...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_REGULAR_WAIT_FOR_COMBUF test with in order queue failed
command_buffer_wait_for_command_buffer FAILED
command_buffer_wait_for_sec_command_buffer...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_COMBUF_WAIT_FOR_SEC_COMBUF test with in order queue failed
command_buffer_wait_for_sec_command_buffer FAILED
return_event_callback...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_EVENT_CALLBACK test with in order queue failed
return_event_callback FAILED
clwaitforevents_single...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_CLWAITFOREVENTS_SINGLE test with in order queue failed
clwaitforevents_single FAILED
clwaitforevents...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_CLWAITFOREVENTS test with in order queue failed
clwaitforevents FAILED
command_buffer_wait_for_regular...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_COMBUF_WAIT_FOR_REGULAR test with in order queue failed
command_buffer_wait_for_regular FAILED
wait_for_sec_queue_event...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::RET_WAIT_FOR_SEC_QUEUE_EVENT test with in order queue failed
wait_for_sec_queue_event FAILED
user_event_wait...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::USER_EVENT_WAIT test with in order queue failed
user_event_wait FAILED
user_events_wait...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::USER_EVENTS_WAIT test with in order queue failed
user_events_wait FAILED
user_event_callback...
Device does not support 'cl_khr_command_buffer'. Skipping the test.
EventMode::USER_EVENT_CALLBACK test with in order queue failed
user_event_callback FAILED
```

Reply:
you're obviously right. With doubled test there is a slight concern when it may be skipped. Since out-of-order command queue is optional feature of OpenCL I rearranged general steps to follow this schema:

-perform test with out-of-order command queue
    -if test failed return TEST_FAIL  
    -if test skipped log the message but continue the test
-perform test with in-order command queue test
    -if test failed return TEST_FAIL
    -if test skipped log the message and return TEST_SKIP